### PR TITLE
Revise LLaMA ingredients testing

### PR DESCRIPTION
**Pull request by amkatrutsa**
Address #114 #115 #116 issues

### CODE REVIEW

File: wrappers/python/nntile/model/llama.py [hunk 1]
```diff
@@ -88,6 +89,23 @@ def from_torch(torch_llama,
                                     config.hidden_size_tile,
                                     config.hidden_size_tile,
                                     next_tag)
+        elif config.dtype == "bf16":
+            embed_layer, next_tag = Embedding.generate_simple(
+                                    x_value, Tensor_bf16, 0,
```
Comment (Muxas) [inline wrappers/python/nntile/model/llama.py:94]: make all `if` statements only set variable `tensor_type` to one of `Tensor_fp32`, ... Create embedding layer after the new shortened `if` clause.

Comment (Muxas) [reply wrappers/python/nntile/model/llama.py:94]: Alternatively, a `dict` to map string representation to a proper Tensor type can be used

File: wrappers/python/tests/model/test_llama_decoder.py [hunk 2]
```diff
@@ -180,69 +135,77 @@ def generate_inputs(params: LlamaDecoderTestParams):
...
+])
 class TestLlamaMLP:
-    def test_from_torch_and_to_torch(
-        self, starpu_simple, torch_rng, params: LlamaDecoderTestParams
-    ):
-        torch_layer, nntile_layer, _, _, _, _ = generate_inputs(params)
+    def test_from_torch_and_to_torch(self, starpu_simple, torch_rng,
+                                     params: LlamaDecoderTestParams,
+                                     dtype: str):
+        torch_layer, nntile_layer, _, _, _, _ = generate_inputs(params, dtype)
```
Comment (Muxas) [inline wrappers/python/tests/model/test_llama_decoder.py:151]: it is possible to shorten `..., _, _, _, _ =` into `..., *_ =`

File: wrappers/python/tests/model/test_llama_mlp.py [hunk 3]
```diff
@@ -55,79 +58,31 @@ class LlamaMLPTestParams:
...
+    n_batch=4,
+    n_batch_tile=1)
 
+single_tile = LlamaMLPTestParams(
+    hidden_size=128,
+    hidden_size_tile=128,
+    intermediate_size=64,
+    intermediate_size_tile=64,
+    n_batch=3,
+    n_batch_tile=3,
```
Comment (Muxas) [inline wrappers/python/tests/model/test_llama_mlp.py:81]: make `seq_len` also divided into several tiles

File: wrappers/python/tests/model/test_llama_decoder.py [hunk 4]
```diff
@@ -55,79 +58,30 @@ class LlamaDecoderTestParams:
...
+    intermediate_size_tile=64,
+    n_batch=3,
+    n_batch_tile=3)
+
+multiple_tiles = LlamaDecoderTestParams(
+    hidden_size=128,
+    hidden_size_tile=32,
+    intermediate_size=64,
+    intermediate_size_tile=16,
+    n_batch=4,
```
Comment (Muxas) [inline wrappers/python/tests/model/test_llama_decoder.py:80]: Add `seq_len` parameter and divide it into tiles

**Review by Muxas — CHANGES_REQUESTED** [review]

**Comment by Muxas** [comment]
I think it is possible to even further simplify testing by checking list of parameters of both `torch_layer` and `torch_layer_other` instead of specifying them explicitly by names

File: wrappers/python/tests/model/test_llama.py [hunk 5]
```diff
@@ -167,73 +184,61 @@ def generate_inputs(params: LlamaTestParams):
...
-                torch_model.layers[i].self_attn.k_proj.weight.detach().numpy(),
-                torch_model_other.layers[i].self_attn.k_proj.weight.detach().numpy(),
-                **dtype2tol[params.dtype]
-            )
+        assert_close_by_frobnorm(
+            torch_model.norm.weight.detach().numpy(),
+            torch_model_other.norm.weight.detach().numpy(),
+            **dtype2tol[dtype]
+        )
+        for i in range(num_hidden_layers):
```
Comment (Muxas) [inline wrappers/python/tests/model/test_llama.py:219]: Embed layer is checked, transformer blocks are checked, but the last rmsnorm is not checked fully. At least, bias `torch_model.norm.bias` is not checked here.

Comment (Muxas) [reply wrappers/python/tests/model/test_llama.py:219]: Moreover, biases are not checked at all.

File: wrappers/python/tests/model/test_llama_decoder.py [hunk 6]
```diff
@@ -180,69 +138,43 @@ def generate_inputs(params: LlamaDecoderTestParams):
...
-            **dtype2tol[params.dtype]
-        )
-
-    def test_forward(
-        self, starpu_simple, torch_rng, params: LlamaDecoderTestParams
-    ):
+        for parameter_name, _ in torch_layer.named_parameters():
+            split_name = parameter_name.split(".")
+            if len(split_name) == 2:
+                assert_close_by_frobnorm(
```
Comment (Muxas) [inline wrappers/python/tests/model/test_llama_decoder.py:161]: biases are not checked

**Review by Muxas — CHANGES_REQUESTED** [review]
Biases are not checked at all. Only weights.

File: wrappers/python/tests/model/test_llama.py (hunk 5 above)
Comment (amkatrutsa) [reply wrappers/python/tests/model/test_llama.py:220]: The last norm is checked after the for-loop over the decoder blocks. RMSNorm has only a weight parameter.

File: wrappers/python/tests/model/test_llama_decoder.py [hunk 7]
```diff
@@ -245,51 +214,17 @@ def test_forward_backward(self, starpu_simple, torch_rng,
...
+                    torch_layer.__getattr__(split_name[0]).weight.grad.detach().numpy(),
+                    torch_layer_other.__getattr__(split_name[0]).weight.grad.detach().numpy(),
+                    **dtype2tol[dtype]
+                )
+            elif len(split_name) == 3:
+                assert_close_by_frobnorm(
+                    torch_layer.__getattr__(split_name[0]).__getattr__(split_name[1]).weight.grad.detach().numpy(),
+                    torch_layer_other.__getattr__(split_name[0]).__getattr__(split_name[1]).weight.grad.detach().numpy(),
+                    **dtype2tol[dtype]
+                )
```
Comment (daskol) [inline wrappers/python/tests/model/test_llama_decoder.py:254]: You should never directly call `__getattr__`. Python has quite complex attribute lookup mechanism and `__getattr__` is a single piece of that machinery. Thus `getattr(obj, attr_name, default_value)` should be used instead.

In addition, if you need access multiple attributes like `x.y.z` then you can use `operator.attrgetter` as follows.
```python
from operator import attrgetter
x = ...  # Complex object.
z = attrgetter('y.z')(x)
```
You can likely meet it in sorting an iterable of complex objects.
```python
sorted(xs, key=attrgetter('y.z'))  # Use `x.y.z` as a sorting key.
```

**Review by daskol — CHANGES_REQUESTED** [review]

File: wrappers/python/tests/model/test_llama_decoder.py (hunk 7 above)
Comment (Muxas) [reply wrappers/python/tests/model/test_llama_decoder.py]: No `__getattr__` is used now. Instead, we iterate over `named_parameters()`.

**Review by Muxas — APPROVED** [review]

**Pull request merged**